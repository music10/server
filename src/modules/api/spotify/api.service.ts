import { HttpService, Injectable } from '@nestjs/common';
import { MusicApi, Track } from '../../../interfaces';

/**
 * Service for Spotify
 * @implements MusicApi
 * @interface MusicApi
 */
@Injectable()
export class ApiService implements MusicApi {
  /**
   * Token for Spotify API
   * Random value by default
   * @private
   */
  private headers = {
    Authorization:
      'Bearer BQAOGPdvZEAULEm9VsQdtHaxyzHn1i2niRQNJG5Uw2-txmlpOTsM6PfSbDjrKRBFiQbS7UciS_Ajhm7CZUs',
  };

  /**
   * SpotifyApiService constructor
   */
  constructor(private httpService: HttpService) {
    this.httpService.axiosRef.interceptors.request.use((config) => ({
      ...config,
      headers: { ...this.headers, ...config.headers },
    }));

    this.httpService.axiosRef.interceptors.response.use(
      (value) => value,
      async (error) => {
        if (
          error.response?.data?.error?.message === 'The access token expired'
        ) {
          return await this.httpService
            .post(
              'https://accounts.spotify.com/api/token',
              'grant_type=client_credentials',
              {
                headers: {
                  Authorization: `Basic ${process.env.SPOTIFY_AUTH_TOKEN}`,
                },
              },
            )
            .toPromise()
            .then(({ data }) => {
              const newToken = `Bearer ${data['access_token']}`;
              this.headers.Authorization = newToken;
              const newConfig = error.config;
              newConfig.headers.Authorization = newToken;
              return this.httpService.axiosRef.request(newConfig);
            });
        }
        return Promise.reject(error);
      },
    );
  }

  /**
   * Get playlists
   */
  async getPlaylists() {
    return this.httpService
      .get('https://music.dergunov.net/allowed-playlists.json')
      .toPromise()
      .then(({ data }) => data);
  }
  /**
   * Search playlists by query-string
   * @param query
   * @return
   */
  async searchPlaylists(query) {
    return (await this.getPlaylists()).filter((playlist) =>
      new RegExp(query, 'ig').test(playlist.name),
    );
  }

  /**
   * Get playlist by id
   * @param playlistId - playlist id
   * @return playlist
   */
  async getPlaylistById(playlistId) {
    return this.httpService
      .get(`/playlists/${playlistId}`, {
        params: {
          fields: 'id,name',
        },
      })
      .toPromise()
      .then(({ data }) => ({
        name: data.name,
        id: data.id,
        getTracks: () => this.getTracksByPlaylistId(playlistId),
      }));
  }

  /**
   * Get tracks by playlist id
   * @param playlistId - playlist id
   * @return tracks
   */ async getTracksByPlaylistId(playlistId) {
    const tracks = [];
    return new Promise<Track[]>((resolve) => {
      const responseHandler = ({ data }) => {
        tracks.push(
          ...data.items.map(({ track }) => ({
            name: track.name,
            id: track.id,
            author: track.artists.map((artist) => artist.name).join(', '),
            mp3: track.preview_url,
          })),
        );

        if (data.next) {
          this.httpService.get(data.next).toPromise().then(responseHandler);
        } else {
          resolve(tracks);
        }
      };

      this.httpService
        .get(`/playlists/${playlistId}/tracks`, {
          params: {
            fields: 'next,items(track(id,artists(name),name,preview_url))',
          },
        })
        .toPromise()
        .then(responseHandler);
    });
  }

  /**
   * Get track by id
   * @param trackId
   * @return track
   */
  async getTrackById(trackId) {
    return this.httpService
      .get(`/tracks/${trackId}`, {
        params: {
          fields: 'id,name,artists(name),preview_url',
        },
      })
      .toPromise()
      .then(({ data }) => ({
        author: data.artists.map((artist) => artist.name).join(', '),
        mp3: data.preview_url,
        name: data.name,
        id: data.id,
      }));
  }
}
