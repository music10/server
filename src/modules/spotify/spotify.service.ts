import { HttpService, Injectable } from '@nestjs/common';
import { Playlist, Track } from '../../interfaces';

const SPOTIFY_OWNER_ID = 'spotify';

/**
 * Service for Spotify Api
 */
@Injectable()
export class SpotifyService {
  /**
   * Token for Spotify API
   * Random value by default
   * @private
   */
  private headers = {
    Authorization:
      'Bearer BQAOGPdvZEAULEm9VsQdtHaxyzHn1i2niRQNJG5Uw2-txmlpOTsM6PfSbDjrKRBFiQbS7UciS_Ajhm7CZUs',
  };

  /**
   * SpotifyApiService constructor
   */
  constructor(private httpService: HttpService) {
    this.httpService.axiosRef.interceptors.request.use((config) => ({
      ...config,
      headers: { ...this.headers, ...config.headers },
    }));

    this.httpService.axiosRef.interceptors.response.use(
      (value) => value,
      async (error) => {
        if (
          error.response?.data?.error?.message === 'The access token expired'
        ) {
          return await this.httpService
            .post(
              'https://accounts.spotify.com/api/token',
              'grant_type=client_credentials',
              {
                headers: {
                  Authorization: `Basic ${process.env.SPOTIFY_AUTH_TOKEN}`,
                },
              },
            )
            .toPromise()
            .then(({ data }) => {
              const newToken = `Bearer ${data['access_token']}`;
              this.headers.Authorization = newToken;
              const newConfig = error.config;
              newConfig.headers.Authorization = newToken;
              return this.httpService.axiosRef.request(newConfig);
            });
        }
        return Promise.reject(error);
      },
    );
  }

  /**
   * Get playlists
   */
  async getCherryPickPlaylists(): Promise<Playlist[]> {
    return this.httpService
      .get('https://music.dergunov.net/allowed-playlists.json')
      .toPromise()
      .then(({ data }) => data);
  }

  /**
   * Search playlists by query-string
   * @param query
   * @return
   */
  async searchPlaylists(query): Promise<Playlist[]> {
    return this.httpService
      .get(`/search`, {
        params: {
          q: query,
          type: 'playlist',
        },
      })
      .toPromise()
      .then(({ data }) =>
        data.playlists.items
          .filter((playlist) => playlist.tracks.total > 40)
          .slice(0, 20)
          .map(({ id, name, images }) => ({
            id,
            name,
            cover: images[0].url,
          })),
      );
  }

  /**
   * Search playlists by artist via query-string
   * @param query
   * @return
   */
  async searchPlaylistsByArtist(query): Promise<Playlist[]> {
    return this.httpService
      .get(`/search`, {
        params: {
          q: `This is ${query}`,
          type: 'playlist',
        },
      })
      .toPromise()
      .then(({ data }) =>
        data.playlists.items
          .filter(
            ({ tracks, owner }) =>
              tracks.total > 40 && owner.id === SPOTIFY_OWNER_ID,
          )
          .slice(0, 20)
          .map(({ id, name, images }) => ({
            id,
            name,
            cover: images[0].url,
          })),
      );
  }

  /**
   * Get playlist by id
   * @param playlistId - playlist id
   * @return playlist
   */
  async getPlaylistById(playlistId): Promise<Playlist> {
    return this.httpService
      .get(`/playlists/${playlistId}`, {
        params: {
          fields: 'id,name,images,href',
        },
      })
      .toPromise()
      .then(({ data }) => ({
        name: data.name,
        id: data.id,
        cover: data.images[0].url,
        getTracks: () => this.getTracksByPlaylistId(playlistId),
      }));
  }

  /**
   * Get tracks by playlist id
   * @param playlistId - playlist id
   * @return tracks
   */ async getTracksByPlaylistId(playlistId): Promise<Track[]> {
    const tracks = [];
    return new Promise<Track[]>((resolve) => {
      const responseHandler = ({ data }) => {
        tracks.push(
          ...data.items
            .filter(({ track }) => track?.preview_url)
            .map(({ track }) => ({
              name: track.name,
              id: track.id,
              album: data.album.name,
              artist: track.artists.map((artist) => artist.name).join(', '),
              mp3: track.preview_url,
            })),
        );

        if (data.next) {
          this.httpService.get(data.next).toPromise().then(responseHandler);
        } else {
          resolve(tracks);
        }
      };

      this.httpService
        .get(`/playlists/${playlistId}/tracks`, {
          params: {
            fields:
              'next,items(track(id,artists(name),name,preview_url,album(name))',
          },
        })
        .toPromise()
        .then(responseHandler);
    });
  }

  /**
   * Get track by id
   * @param trackId
   * @return track
   */
  async getTrackById(trackId): Promise<Track> {
    return this.httpService
      .get(`/tracks/${trackId}`, {
        params: {
          fields: 'id,name,artists(name),album(name),preview_url',
        },
      })
      .toPromise()
      .then(({ data }) => ({
        id: data.id,
        name: data.name,
        album: data.album.name,
        artist: data.artists.map((artist) => artist.name).join(', '),
        mp3: data.preview_url,
      }));
  }
}
