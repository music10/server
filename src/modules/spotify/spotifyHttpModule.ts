import { HttpModule } from '@nestjs/axios';

/**
 * Http module for Spotify API
 */
export const SpotifyHttpModule = HttpModule.registerAsync({
  useFactory: () => ({
    timeout: 5000,
    maxRedirects: 5,
    baseURL: 'https://api.spotify.com/v1',
    headers: {
      'accept-language': 'ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7',
    },
  }),
});
