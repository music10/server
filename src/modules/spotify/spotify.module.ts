import { Module } from '@nestjs/common';
import { SpotifyHttpModule } from './spotifyHttpModule';
import { SpotifyService } from './spotify.service';

@Module({
  imports: [SpotifyHttpModule],
  providers: [SpotifyService],
  exports: [SpotifyService],
})
export class SpotifyModule {}
