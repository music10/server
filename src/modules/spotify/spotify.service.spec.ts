import { Test, TestingModule } from '@nestjs/testing';
import { SpotifyService } from './spotify.service';
import { SpotifyHttpModule } from './spotifyHttpModule';

describe('SpotifyApiService', () => {
  let service: SpotifyService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [SpotifyHttpModule],
      providers: [SpotifyService],
    }).compile();

    service = module.get<SpotifyService>(SpotifyService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should get playlists', async () => {
    expect(await service.getCherryPickPlaylists()).toMatchSnapshot();
  });

  it('should search playlists', async () => {
    expect(await service.searchPlaylists('русский')).toEqual(
      expect.arrayContaining([
        {
          id: expect.any(String),
          name: expect.any(String),
          cover: expect.any(String),
        },
      ]),
    );
  });

  it('should search playlists by artist', async () => {
    expect(await service.searchPlaylistsByArtist('Ice Cube')).toHaveLength(1);
  });

  it('should get playlist by id', async () => {
    const result = await service.getPlaylistById('37i9dQZF1DXathzFlVFFpY');
    expect(result).toStrictEqual({
      cover: 'https://i.scdn.co/image/ab67706f000000034647a679d5cc388b5c58fa13',
      id: '37i9dQZF1DXathzFlVFFpY',
      name: 'Хиты русского рэпа',
      getTracks: expect.any(Function),
    });
    expect((await result.getTracks()).length).toBeGreaterThan(10);
  });

  it('should get tracks by playlist id', async () => {
    const result = await service.getTracksByPlaylistId(
      '37i9dQZF1DXathzFlVFFpY',
    );
    expect(result.length).toBeGreaterThan(10);
    expect(result[0]).toStrictEqual({
      artist: 'Баста',
      mp3:
        'https://p.scdn.co/mp3-preview/82baf3251c6681495a3cb5b2c9b476b4e51f2070?cid=6e32a60ae68b408596f337136300880c',
      id: '1xIDRR91yrYa3LvYWkOxxz',
      name: 'Моя игра',
      album: 'Баста 1',
    });
  });

  it('should get track by id', async () => {
    const result = await service.getTrackById('1xIDRR91yrYa3LvYWkOxxz');
    expect(result).toStrictEqual({
      artist: 'Баста',
      mp3:
        'https://p.scdn.co/mp3-preview/82baf3251c6681495a3cb5b2c9b476b4e51f2070?cid=6e32a60ae68b408596f337136300880c',
      id: '1xIDRR91yrYa3LvYWkOxxz',
      name: 'Моя игра',
      album: 'Баста 1',
    });
  });
});
