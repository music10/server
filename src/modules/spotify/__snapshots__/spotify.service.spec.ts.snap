// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifyApiService should get playlists 1`] = `
Array [
  Object {
    "cover": "https://music.dergunov.net/images/world.png",
    "id": "5SHptL6bzOG1PmWgJxx2Ps",
    "name": "Топ по миру",
  },
  Object {
    "cover": "https://music.dergunov.net/images/russia.png",
    "id": "5I4WLaJYpw1pPcRXwUQ9XV",
    "name": "Топ по России",
  },
  Object {
    "cover": "https://music.dergunov.net/images/new.png",
    "id": "37i9dQZF1DWYc8GECSzPrb",
    "name": "Новинки недели",
  },
  Object {
    "cover": "https://music.dergunov.net/images/tiktok.png",
    "id": "56ra24CW6btsEQEA3UEzqD",
    "name": "Топ тик-тока",
  },
  Object {
    "cover": "https://music.dergunov.net/images/hiphop.png",
    "id": "37i9dQZF1DX93L1kSOomqn",
    "name": "Русский хип-хоп",
  },
  Object {
    "cover": "https://music.dergunov.net/images/rock.png",
    "id": "37i9dQZF1DX4uWsCu3SlsH",
    "name": "Рок сегодня",
  },
]
`;
